client.go

```diff
@@ -248,20 +252,24 @@ func (r remoteProcess) Resize(ctx context.Context, rows, cols uint16) error {
 	return r.conn.Write(ctx, websocket.MessageBinary, payload)
 }
 
-func (r remoteProcess) Wait() error {
-	select {
-	case err := <-r.done:
-		return err
-	case <-r.ctx.Done():
-		return r.ctx.Err()
+func (r *remoteProcess) Wait() error {
+	<-r.done
+	if r.readErr != nil {
+		return r.readErr
+	}
+	// when listen() closes r.done, either there must be a read error
+	// or exitCode is set non-nil, so it's safe to dereference the pointer
+	// here
+	if *r.exitCode != 0 {
+		return ExitError{Code: *r.exitCode}
 	}
+	return nil
 }
 
-func (r remoteProcess) Close() error {
-	err := r.conn.Close(websocket.StatusNormalClosure, "")
-	err1 := r.stderr.w.Close()
-	err2 := r.stdout.w.Close()
-	return joinErrs(err, err1, err2)
+func (r *remoteProcess) Close() error {
+	r.cancelListen()
+	<-r.done
+	return joinErrs(r.closeErr, r.stdoutErr, r.stderrErr)
```

Comment:
This is mostly just to make sure I follow, will calling `Close` block on `r.conn.Read`?  I was trying to see if `conn.Read` aborts when the context is canceled but it seems to wait for a read or possibly a timeout so I am confused how it unblocks unless canceling also triggers the read timeout or something.

Reply:
Yeah, it will time out.  I'm adding some testing here and I've noticed a problem, which is that a read timeout will also close the websocket, so our call to close then throws an error.  I'll fix it up.

I'm also noticing that we can get stuck copying stdout and stderr, so I'll also fix that up.